Documentation of the testing framework (#178)
quarkiverse/quarkus-github-app: docs/modules/ROOT/pages/testing.adoc

= Testing

include::./includes/attributes.adoc[]

[[manual-vs-automated]]
== Manual testing vs. automated testing

The easiest way to start testing your GitHub application is probably
to create a playground project on GitHub and to run the application locally through Smee.io,

> as explained in xref:register-github-app.adoc[register the application].

Reviewer: ```suggestion
as explained in xref:register-github-app.adoc[Register a GitHub App].
```

Let's continue with the example from xref:create-github-app.adoc[Create a GitHub App]:
open an issue in your playground repository,
go to the replay UI to copy the event payload,
then put that payload in a file in your project: `src/test/resources/issue-opened.json`.

== Create a test class

Add the following class to `src/test/java`, in a package of your choosing:

[source, java]
----
@QuarkusTest
@GitHubAppTest
class CreateCommentTest {
    @Test
    void testIssueOpened() {
        GitHubAppTesting.when() // <1>
                .payloadFromClasspath("/issue-opened.json") // <2>
                .event(GHEvent.ISSUES) // <3>
                .then().github(mocks -> { // <4>
                    Mockito.verify(mocks.issue(750705278)) // <5>
                            .comment("Hello from my GitHub App"); // <6>
                });
    }
}
----
<1> Use `GitHubAppTesting.when()` to start simulating an event.
<2> Define the payload of the simulated event by pointing to a file in the classpath.
<3> Define the type of the simulated event.
<4> Use `.then().github(mocks -> ...)` to perform assertions on GitHub objects involved in the event handling.

> configure mocks.

Reviewer: There's something weird here.
